x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/QueryRulesIndexService.java

```diff
@@ -0,0 +1,362 @@
+     *
+     * TODO add total number of rules per ruleset - We can add this when implementing the List command.
+     */
+    public void listQueryRulesets(int from, int size, ActionListener<QueryRulesetResult> listener) {
+        try {
+            final SearchSourceBuilder source = new SearchSourceBuilder().from(from)
+                .size(size)
+                .query(new MatchAllQueryBuilder())
+                .docValueField(QueryRuleset.ID_FIELD.getPreferredName())
+                .storedFields(Collections.singletonList("_none_"))
+                .sort(QueryRuleset.ID_FIELD.getPreferredName(), SortOrder.ASC);
+            final SearchRequest req = new SearchRequest(QUERY_RULES_ALIAS_NAME).source(source);
+            clientWithOrigin.search(req, new ActionListener<>() {
+                @Override
+                public void onResponse(SearchResponse searchResponse) {
+                    listener.onResponse(mapSearchResponseToQueryRulesetList(searchResponse));
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    if (e instanceof IndexNotFoundException) {
+                        listener.onResponse(new QueryRulesetResult(Collections.emptyList(), 0L));
+                        return;
+                    }
+                    listener.onFailure(e);
+                }
+            });
+        } catch (Exception e) {
+            listener.onFailure(e);
+        }
+    }
+
+    private static QueryRulesetResult mapSearchResponseToQueryRulesetList(SearchResponse response) {
+        final List<String> rulesetIds = Arrays.stream(response.getHits().getHits())
+            .map(hit -> (String) hit.getDocumentFields().get(QueryRuleset.ID_FIELD.getPreferredName()).getValue())
+            .toList();
+        return new QueryRulesetResult(rulesetIds, (int) response.getHits().getTotalHits().value);
+    }
+
+    static QueryRuleset parseQueryRulesetBinaryWithVersion(StreamInput in) throws IOException {
+        TransportVersion version = TransportVersion.readVersion(in);
+        assert version.onOrBefore(TransportVersion.CURRENT) : version + " >= " + TransportVersion.CURRENT;
+        in.setTransportVersion(version);
+        return new QueryRuleset(in);
+    }
+
+    static QueryRule parseQueryRuleBinaryWithVersion(StreamInput in) throws IOException {
+        TransportVersion version = TransportVersion.readVersion(in);
+        assert version.onOrBefore(TransportVersion.CURRENT) : version + " >= " + TransportVersion.CURRENT;
+        in.setTransportVersion(version);
+        return new QueryRule(in);
+    }
+
+    static void writeQueryRulesetBinaryWithVersion(QueryRuleset queryRuleset, OutputStream os, TransportVersion minTransportVersion)
+        throws IOException {
+        // do not close the output
+        os = Streams.noCloseStream(os);
+        TransportVersion.writeVersion(minTransportVersion, new OutputStreamStreamOutput(os));
```

Comment:
What are the pros/cons of not doing binary serialization? It's space vs. convenience right? Anything I'm missing? I'd be OK if we just want to store everything as text.

Reply:
I think index readability should take precedence, specially this early in development. Remember when we couldn't create search apps unless we used PUT because of the binary serialization? Ah, good ol' times... 😆 

I think @jimczi had his reasons for doing binary serialization, but I don't think I really got to understand the need for using them.

Reply:
I choose the binary serialisation because that's where we handle versioning currently. We don't do it consistently in our rest layer but we should probably fix that. The recommended path now is to use versioned APIs so I agree with Carlos we should try the rest serialisation for this new one. In general relying on _source directly is a good approach to retrieve objects from a system index. The mapping should not be used and we can simply declare a disabled object to consider the resource as a json blob.